// @framer/northam.animated-vertical-menu/code/App.tsx
import { Data, animate, Override, Animatable } from "framer"
import { Curve } from "framer/types/src/animation/Animators/BezierAnimator";

const data = Data({
    menuOpen: false,
    top: Animatable(-140),
    rotation: Animatable(0)
})

export const menuOverride: Override = () => {
    return {
        top: data.top,
    }
}

export const arrowOverride: Override = props => {
    data.rotation.set(props.rotation)
    return { 
        rotation: data.rotation,
    }
  }

const springOptions = { tension: 500, friction: 50 };

const bezierCurve = [.31,.03,.23,.99] as Curve;
const duration = 0.5;
const bezierOptions = { curve: bezierCurve, duration };

export const toggleMenu: Override = () => {
    return {
        onTap: () => {
            data.menuOpen = !data.menuOpen;
            if (data.menuOpen) {
                // Open menu
                animate.bezier(data.top, 30, bezierOptions);
                animate.spring(data.rotation, 180, springOptions);
            } else {
                // Close menu
                animate.bezier(data.top, -140, bezierOptions);
                animate.spring(data.rotation, 0, springOptions);
            }
        }
    }
}
