// @framer/kasimir.iphone-x-kit/code/Status Bar/Status_Bar.tsx
import * as React from "react"
import { DeprecatedFrame as Frame, PropertyControls, ControlType } from "framer"

type Appearance = "dark" | "light"
type Mode = "transparent" | "#4CD964" | "#007AFF" | "#FF3B30"
type Location = "none" | "background" | "locating"

interface Props {
    appearance: Appearance
    mode: Mode
    time: string
    twelveHour: boolean
    liveTime: boolean
    location: Location
    back: string
    batteryLevel: number
    signalLevel: number
    wifiLevel: number
    width: number
    height: number
}

interface State {
    time: string
}

export class StatusBar extends React.Component<Partial<Props>, State> {
    static defaultProps = {
        appearance: "light",
        mode: "transparent",
        liveTime: true,
        back: "",
        twelveHour: false,
        time: "18.00",
        location: "none",
        signalLevel: 100,
        wifiLevel: 100,
        batteryLevel: 100,
        levels: null,
        width: 375,
        height: 44,
    }

    static propertyControls: PropertyControls<Props> = {
        appearance: {
            type: ControlType.Enum,
            options: ["dark", "light"],
            optionTitles: ["Dark", "Light"],
            title: "Appearance",
        },
        mode: {
            type: ControlType.Enum,
            options: ["transparent", "#4CD964", "#007AFF", "#FF3B30"],
            optionTitles: [
                "None",
                "On a call",
                "App actively using your location or Personal Hotspot is on",
                "Recording",
            ],
            title: "Task",
        },
        liveTime: {
            type: ControlType.Boolean,
            disabledTitle: "No",
            enabledTitle: "Yes",
            title: "Real time",
        },
        twelveHour: {
            type: ControlType.Boolean,
            disabledTitle: "24h",
            enabledTitle: "12h",
            title: "↳ Time format",
            hidden(props) {
                return props.liveTime === false
            },
        },
        time: {
            type: ControlType.String,
            placeholder: "18:00",
            title: "↳ Time",
            hidden(props) {
                return props.liveTime
            },
        },
        location: {
            type: ControlType.Enum,
            options: ["none", "background", "locating"],
            optionTitles: [
                "No Location",
                "Background Location",
                "Receiving Location",
            ],
            title: "Location",
        },
        back: {
            type: ControlType.String,
            placeholder: "Settings",
            title: "Back to",
        },
        signalLevel: {
            type: ControlType.Number,
            min: 0,
            max: 100,
            title: "Signal",
        },
        wifiLevel: {
            type: ControlType.Number,
            min: 0,
            max: 100,
            title: "Wifi",
        },
        batteryLevel: {
            type: ControlType.Number,
            min: 0,
            max: 100,
            title: "Battery",
        },
    }

    time: "0"

    timer: number

    state = {
        time: "18:00",
    }

    componentWillMount() {
        this.updateTime()

        if (this.props.liveTime) {
            this.timer = window.setInterval(() => {
                this.updateTime()
            }, 1000)
        }
    }

    componentWillUnmount() {
        window.clearInterval(this.timer)
    }

    updateTime = () => {
        let time = ""

        if (this.props.liveTime) {
            const date = new Date()
            var hours = date.getHours()

            if (this.props.twelveHour) {
                hours = hours % 12
                hours = hours ? hours : 12

                time = hours + ":" + ("0" + date.getMinutes()).slice(-2)
            } else {
                time =
                    ("0" + hours).slice(-2) +
                    ":" +
                    ("0" + date.getMinutes()).slice(-2)
            }
        } else {
            time = this.props.time
        }

        if (this.time !== time) {
            this.time = time

            this.setState({ time })
        }
    }

    render() {
        const {
            batteryLevel,
            signalLevel,
            wifiLevel,
            appearance,
            location,
            back,
            mode,
            width,
            height,
        } = this.props

        if (width >= 800) {
            return <div />
        }

        return (
            <div
                style={{
                    display: "flex",
                    justifyContent: "space-between",
                    alignItems: "center",
                    width: "100%",
                    top: 0,
                    left: 0,
                    fontSize: 12,
                    color: appearance === "dark" ? "black" : "white",
                }}
            >
                <div
                    style={{
                        height: "44px",
                        width: "100%",
                        display: "flex",
                        flexDirection: "row",
                        justifyContent: "space-between",
                        alignItems: "center",
                    }}
                >
                    <div
                        style={{
                            display: "flex",
                            height: "44px",
                            width: "90px",
                            flexDirection: "column",
                            alignItems: "center",
                            justifyContent: "center",
                        }}
                    >
                        <div
                            style={{
                                display: "flex",
                                alignItems: "center",
                                justifyContent: "center",
                                textAlign: "center",
                                fontFamily:
                                    "SF-Pro-Text, system-ui, -apple-system",
                                fontSize: "15px",
                                letterSpacing: "-16",
                                width: "54px",
                                lineHeight: "21px",
                                fontWeight: 600,
                                background: mode,
                                color:
                                    mode != "transparent"
                                        ? "#ffffff"
                                        : "inherit",
                                borderRadius: "20px",
                            }}
                        >
                            {this.state.time !== "" ? this.state.time : "18:00"}
                            <div
                                style={{
                                    height: "16px",
                                    display:
                                        mode === "transparent" &&
                                        location !== "none"
                                            ? "block"
                                            : "none",
                                }}
                                dangerouslySetInnerHTML={{
                                    __html: locationSVG(appearance, location),
                                }}
                            />
                        </div>
                        <div
                            style={{
                                alignSelf: "flex-start",
                                display: back != "" ? "flex" : "none",
                                marginLeft: "10px",
                                fontWeight: 400,
                                fontFamily:
                                    "SF-Pro-Text, system-ui, -apple-system",
                                fontSize: "12px",
                            }}
                        >
                            <div
                                style={{
                                    height: "12px",
                                    width: "12px",
                                }}
                                dangerouslySetInnerHTML={{
                                    __html: backSVG(appearance),
                                }}
                            />
                            {back}
                        </div>
                    </div>
                    <div
                        style={{
                            display: "flex",
                            alignItems: "center",
                            justifyContent: "center",
                            height: "44px",
                            width: "95px",
                        }}
                    >
                        <div
                            style={{
                                width: "20px",
                                textAlign: "center",
                            }}
                            dangerouslySetInnerHTML={{
                                __html: signalSVG(appearance, signalLevel),
                            }}
                        />
                        <div
                            style={{
                                width: "20px",
                                textAlign: "center",
                            }}
                            dangerouslySetInnerHTML={{
                                __html: wifiSVG(appearance, wifiLevel),
                            }}
                        />
                        <div
                            style={{
                                width: "27px",
                                textAlign: "right",
                            }}
                            dangerouslySetInnerHTML={{
                                __html: batterySVG(appearance, batteryLevel),
                            }}
                        />
                    </div>
                </div>
            </div>
        )
    }
}

function backSVG(appearance: Appearance) {
    const color = appearance === "dark" ? "black" : "white"

    return `<svg xmlns="http://www.w3.org/2000/svg" width="6px" height="7px" viewBox="0 0 7 8">
        <path fill="${color}" class="c" d="M0,4,7,0V8Z"/>
      </svg>`
}

function locationSVG(appearance: Appearance, location: string) {
    const color = appearance === "dark" ? "black" : "white"
    const fill = location === "locating" ? color : "transparent"

    return `<svg width="16px" height="16px" viewBox="0 0 16 16" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
          <g class="b" transform="translate(2.5 4)" fill-rule="nonzero" stroke-align="inside" stroke="${color}" fill="${fill}" stroke-dashoffset="1" stroke-width="1">
            <path class="c" d="M5.955,10.5,10.5,0,0,4.545H5.955Z"/>
          </g>
      </svg>`
}

function batterySVG(appearance: Appearance, batteryLevel: number) {
    const color = appearance === "dark" ? "black" : "white"
    const batteryJuiceColor =
        batteryLevel <= 10 ? "red" : batteryLevel <= 20 ? "#FFCC01" : color

    return `<svg width="25px" height="12px" viewBox="0 0 25 12" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
          <g id="Symbols" stroke="none" fill="none" fill-rule="evenodd">
              <g id="Bars-/-Status-Bar-/-iPhone-X-/-On-Light" transform="translate(-336.000000, -17.000000)">
                  <g id="Battery" transform="translate(336.000000, 17.000000)">
                      <rect id="Border" stroke="${color}" opacity="0.35" x="0.5" y="0.833333333" width="21" height="10.3333333" rx="2.66666675"></rect>
                      <path d="M23,4 L23,8 C23.8047311,7.66122348 24.328038,6.87313328 24.328038,6 C24.328038,5.12686672 23.8047311,4.33877652 23,4" id="Cap" fill="${color}" fill-rule="nonzero" opacity="0.4"></path>
                      <rect id="Capacity" fill="${batteryJuiceColor}" fill-rule="nonzero" x="2" y="2.33333333" height="7.33333333" rx="1.33333337" width="${Math.round(
        (batteryLevel / 100) * 17 + 1
    )}"></rect>
                  </g>
              </g>
          </g>
      </svg>`
}

function signalSVG(appearance: Appearance, signalLevel: number) {
    const color = appearance === "dark" ? "black" : "white"

    return `<svg width="17.5px" height="12px" viewBox="0 0 18 12" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
          <g id="Symbols" stroke="none" fill="none" fill-rule="evenodd">
              <g id="Bars-/-Status-Bar-/-iPhone-X-/-On-Light" transform="translate(-293.000000, -17.000000)" fill-rule="nonzero" fill="${color}">
                <path d="M295.666667,24.3333333 C296.218951,24.3333333 296.666667,24.7810486 296.666667,25.3333333 L296.666667,27.3333333 C296.666667,27.8856181 296.218951,28.3333333 295.666667,28.3333333 L294.666667,28.3333333 C294.114382,28.3333333 293.666667,27.8856181 293.666667,27.3333333 L293.666667,25.3333333 C293.666667,24.7810486 294.114382,24.3333333 294.666667,24.3333333 L295.666667,24.3333333 Z" opacity="${
                    signalLevel != 0 ? 1 : 0.4
                }"></path>
                <path d="M299.333333,22.3333333 L300.333333,22.3333333 C300.885618,22.3333333 301.333333,22.7810486 301.333333,23.3333333 L301.333333,27.3333333 C301.333333,27.8856181 300.885618,28.3333333 300.333333,28.3333333 L299.333333,28.3333333 C298.781049,28.3333333 298.333333,27.8856181 298.333333,27.3333333 L298.333333,23.3333333 C298.333333,22.7810486 298.781049,22.3333333 299.333333,22.3333333 Z" opacity="${
                    signalLevel >= 25 ? 1 : 0.4
                }"></path>
                <path d="M304,20 L305,20 C305.552285,20 306,20.4477153 306,21 L306,27.3333333 C306,27.8856181 305.552285,28.3333333 305,28.3333333 L304,28.3333333 C303.447715,28.3333333 303,27.8856181 303,27.3333333 L303,21 C303,20.4477153 303.447715,20 304,20 Z" opacity="${
                    signalLevel >= 50 ? 1 : 0.4
                }"></path>
                <path d="M308.666667,17.6666667 L309.666667,17.6666667 C310.218951,17.6666667 310.666667,18.1143819 310.666667,18.6666667 L310.666667,27.3333333 C310.666667,27.8856181 310.218951,28.3333333 309.666667,28.3333333 L308.666667,28.3333333 C308.114382,28.3333333 307.666667,27.8856181 307.666667,27.3333333 L307.666667,18.6666667 C307.666667,18.1143819 308.114382,17.6666667 308.666667,17.6666667 Z" opacity="${
                    signalLevel >= 75 ? 1 : 0.4
                }"></path>
              </g>
          </g>
      </svg>`
}

function wifiSVG(appearance: Appearance, wifiLevel: number) {
    const color = appearance === "dark" ? "black" : "white"

    return `<svg width="15px" viewBox="0 0 16 12" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
          <g id="Symbols" fill="none" fill-rule="evenodd">
              <g id="Bars-/-Status-Bar-/-iPhone-X-/-On-Light" transform="translate(-315.000000, -17.000000)" fill="${color}" fill-rule="nonzero">
                <path d="M325.549333,26.188 C325.611087,26.127394 325.645095,26.0439919 325.643328,25.957485 C325.641561,25.8709782 325.604176,25.7890342 325.54,25.731 C324.264428,24.6521162 322.396238,24.6521162 321.120667,25.731 C321.056446,25.7889875 321.019,25.8709057 321.01717,25.957413 C321.01534,26.0439203 321.049289,26.1273487 321.111,26.188 L323.108667,28.2036667 C323.167217,28.2629057 323.247042,28.2962443 323.330333,28.2962443 C323.413624,28.2962443 323.49345,28.2629057 323.552,28.2036667 L325.549333,26.188 Z" opacity="${
                    wifiLevel != 0 ? 1 : 0.4
                }"></path>
                <path d="M323.330333,23.4036667 C324.547823,23.4035921 325.721865,23.8561245 326.624333,24.6733333 C326.746397,24.7893139 326.938678,24.7867994 327.057667,24.6676667 L328.212333,23.501 C328.273141,23.439805 328.30688,23.3567887 328.306003,23.2705239 C328.305126,23.1842592 328.269706,23.101946 328.207667,23.042 C325.459468,20.4856177 321.203532,20.4856177 318.455333,23.042 C318.393256,23.1019453 318.357837,23.1842997 318.357023,23.2705923 C318.356208,23.356885 318.390065,23.4398934 318.451,23.501 L319.605333,24.6676667 C319.724322,24.7867994 319.916603,24.7893139 320.038667,24.6733333 C320.940539,23.8566645 322.113649,23.4041708 323.330333,23.4036667 Z" opacity="${
                    wifiLevel >= 33 ? 1 : 0.4
                }"></path>
                <path d="M323.330333,19.608 C325.546226,19.6080971 327.677386,20.4595347 329.283333,21.9863333 C329.404265,22.1042079 329.597562,22.102721 329.716667,21.983 L330.872667,20.8163333 C330.932975,20.7556112 330.966599,20.6733595 330.966098,20.5877792 C330.965598,20.5021988 330.931014,20.420346 330.87,20.3603333 C326.654902,16.3207858 320.005098,16.3207858 315.79,20.3603333 C315.728941,20.4203015 315.694297,20.5021294 315.693734,20.5877099 C315.693171,20.6732903 315.726736,20.755567 315.787,20.8163333 L316.943333,21.983 C317.062362,22.1029017 317.255808,22.1043897 317.376667,21.9863333 C318.982819,20.4594344 321.114223,19.6079944 323.330333,19.608 Z" opacity="${
                    wifiLevel >= 66 ? 1 : 0.4
                }"></path>
              </g>
          </g>
      </svg>`
}
