// @framer/mapbox.mapbox/code/details/CityLocationMap.tsx
import * as React from "react";
import { PropertyControls, ControlType } from "framer";
import { MapViewportCoordinates } from "./MapTypes";
import { MapboxMapCore, Props as MapProps } from "./MapboxMapCore";
import { Cities } from "./Constants";

export interface Props extends MapProps {
  city: string;
}

///
/// A Framer X component that shows a Mapbox map with UI controls for
/// selecting a city for the map to focus on.
///
export class CityLocationMap extends React.Component<Props> {
  static defaultProps: Partial<Props> = {
    city: Cities.names[0],
    ...MapboxMapCore.defaultProps
  };

  static propertyControls: PropertyControls<Props> = {
    ...MapboxMapCore.propertyControls,
    city: {
      type: ControlType.Enum,
      options: Cities.names,
      title: "City"
    }
  };

  spatialCoordinates(city: string): MapViewportCoordinates {
    return {
      width: this.props.width,
      height: this.props.height,
      bearing: 0,
      pitch: 0,
      zoom: 12,
      longitude: -100,
      latitude: 44,
      ...Cities.coordinates[city]
    };
  }

  render() {
    const { city, ...rest } = this.props;
    const viewport = this.spatialCoordinates(city);

    return <MapboxMapCore {...rest} {...viewport} />;
  }
}
