// @framer/mapbox.mapbox/code/details/CustomLocationMap.tsx
import * as React from "react";
import { PropertyControls, ControlType } from "framer";
import { MapViewportCoordinates } from "./MapTypes";
import { MapboxMapCore, Props as MapProps } from "./MapboxMapCore";

export interface Props extends MapProps {}

const defaultSpatialCoordinates: MapViewportCoordinates = {
  latitude: 41.878113,
  longitude: -87.629799,
  bearing: 0,
  pitch: 0,
  zoom: 12,
  width: 375,
  height: 375
};

///
/// A Framer X component that exposes custom controls
/// to set all positioning properties of a Mapbox map.
///
export class CustomLocationMap extends React.Component<Props> {
  static defaultProps: Partial<Props> = {
    ...MapboxMapCore.defaultProps,
    ...defaultSpatialCoordinates
  };

  static propertyControls: PropertyControls<Props> = {
    ...MapboxMapCore.propertyControls,
    longitude: {
      type: ControlType.Number,
      min: -180,
      max: 180,
      step: 0.001,
      title: "Longitude"
    },
    latitude: {
      type: ControlType.Number,
      min: -89,
      max: 90,
      step: 0.001,
      title: "Latitude"
    },
    zoom: {
      type: ControlType.Number,
      min: 1,
      max: 22,
      title: "Zoom",
      step: 0.1
    },
    bearing: {
      type: ControlType.Number,
      min: -180,
      max: 180,
      defaultValue: 0,
      step: 1,
      title: "Bearing"
    },
    pitch: {
      type: ControlType.Number,
      min: 0,
      max: 60,
      step: 1,
      title: "Pitch"
    }
  };

  render() {
    return <MapboxMapCore {...this.props} />;
  }
}
