// @framer/mapbox.mapbox/code/details/MapSearch.tsx
import * as React from "react";
import { PropertyControls, ControlType } from "framer";
import { Languages, MapControlPositions } from './Constants'
import { Props as ControlProps } from "./MapboxControlComponent"

export interface Props {
    showSearch?: boolean
    language?: string
    placeholder?: string
    position?: ControlProps["position"]
    offset?: number
    dropSearchPin?: boolean
    pin?: React.ReactNodeArray
    limitSearch?: boolean
}

const hidden = (props: Partial<Props>) => {
    return props.showSearch === false
}

export const propertyControls: PropertyControls<Props> = {
    showSearch: {
        type: ControlType.Boolean,
        title: "Search",
        defaultValue: false,
        enabledTitle: "Show",
        disabledTitle: "Hide",
    },
    language: {
        type: ControlType.Enum,
        options: Languages.options,
        optionTitles: Languages.titles,
        title: "Language",
        defaultValue: "en",
        hidden,
    },
    placeholder: {
        type: ControlType.String,
        title: "Placeholder",
        defaultValue: "Search",
        hidden,
    },
    position: {
        type: ControlType.Enum,
        options: MapControlPositions.options,
        optionTitles: MapControlPositions.titles,
        title: "Position",
        defaultValue: "top-center",
        hidden,
    },
    offset: {
        type: ControlType.Number,
        title: "Offset",
        max: 100,
        min: 0,
        defaultValue: 25,
        hidden,
    },
    limitSearch: {
        type: ControlType.Boolean,
        title: "Scope",
        defaultValue: false,
        enabledTitle: "Mapview",
        disabledTitle: "Global",
        hidden,
    },
}

export const defaultProps: Props = {
    showSearch: false,
}
