// @framer/kasimir.iphone-x-kit/code/Home Button (iPhone X)/HomeButton.tsx
import * as React from "react"
import { PropertyControls, ControlType } from "framer"

type Appearance = "dark" | "light"

export class HomeButton extends React.Component<Partial<Props>, State> {
    static defaultProps = {
        appearance: "light",
        width: 375,
        height: 34,
    }

    static propertyControls: PropertyControls<Props> = {
        appearance: {
            type: ControlType.Enum,
            options: ["dark", "light"],
            optionTitles: ["Dark", "Light"],
            title: "Appearance",
        },
    }

    render() {
        const { appearance, width, height } = this.props

        return (
            <div
                style={{
                    display: "flex",
                    justifyContent: "center",
                    alignItems: "center",
                    fontSize: 12,
                    color: appearance === "dark" ? "black" : "white",
                }}
            >
                <div
                    style={{
                        marginTop: "6px",
                    }}
                    dangerouslySetInnerHTML={{
                        __html: homeButtonSVG(appearance, width),
                    }}
                />
            </div>
        )
    }
}

function homeButtonSVG(appearance: Appearance, width: Number) {
    const buttonWidth = width >= 800 ? 210 : 134
    const color = appearance === "dark" ? "black" : "white"

    return `<svg width="${buttonWidth}px" height="5px" viewBox="0 0 ${buttonWidth} 5" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
        <!-- Generator: Sketch 51.3 (57544) - http://www.bohemiancoding.com/sketch -->
        <desc>Created with Sketch.</desc>
        <defs></defs>
        <g id="Symbols" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
            <g id="Bars-/-Home-Indicator-/-On-Dark" transform="translate(-121.000000, -20.000000)" fill="${color}">
                <rect id="Home-Indicator" x="121" y="20" width="${buttonWidth}" height="5" rx="2.5"></rect>
            </g>
        </g>
    </svg>`
}
