// @framer/ismael.get-colors-from-images/_draft/Palette.tsx
import * as React from "react"
import { useState, useEffect } from "react"
import { Frame, addPropertyControls, ControlType } from "framer"
import Vibrant = require("node-vibrant")

const defaultBackgroundColor = "rgba(136, 85, 255, 0.1)"
const defaultColor = "#8855FF"

function rgbToString(rgbArray) {
    return `rgb(${rgbArray[0]}, ${rgbArray[1]}, ${rgbArray[2]})`
}

const baseUrl = "http://127.0.0.1:4567/design/images/"

export function PaletteInstance(props) {
    const [state, setState] = useState({
        isLoading: false,
        error: false,
        palette: null,
    })

    function getColor(image: string) {
        setState({
            isLoading: true,
            error: false,
            palette: null,
        })

        return Vibrant.from(image)
            .getPalette()
            .then(palette => {
                setState({
                    isLoading: false,
                    error: false,
                    palette: palette,
                })
            })
            .catch(err => {
                setState({
                    isLoading: false,
                    error: true,
                    palette: null,
                })
            })
    }

    useEffect(() => {
        if (props.children.length) {
            getColor(baseUrl + props.children[0].props.background.src)
        }
    }, [props.children])

    const style = {
        backgroundColor:
            (state.palette &&
                state.palette[props.profile] &&
                rgbToString(state.palette[props.profile]._rgb)) ||
            defaultBackgroundColor,
        color: defaultColor,
        display: "flex",
        justifyContent: "center",
        alignItems: "center",
    }

    return (
        <Frame
            size={"100%"}
            borderRadius={props.borderRadius}
            overflow={"hidden"}
            style={style}
            {...props}
        >
            {state.isLoading && "Loading…"}
            {props.children.length <= 0 && "Connect to an image frame"}
            {state.error && props.children.length && "Error"}
            {state.palette &&
                !state.palette[props.profile] &&
                "Unavailable color profile"}
        </Frame>
    )
}

Palette.defaultProps = {
    borderRadius: 0,
    profile: "Vibrant",
}

addPropertyControls(PaletteInstance, {
    borderRadius: { type: ControlType.Number, title: "Radius" },
    profile: {
        type: ControlType.Enum,
        options: [
            "LightVibrant",
            "Vibrant",
            "DarkVibrant",
            "LightMuted",
            "Muted",
            "DarkMuted",
        ],
        title: "Color profile",
        hidden(props) {
            return props.image === ""
        },
    },
})

export function Palette(props) {
    const [state, setState] = useState({
        isLoading: false,
        error: false,
        palette: null,
    })

    function getColor(image: string) {
        setState({
            isLoading: true,
            error: false,
            palette: null,
        })

        Vibrant.from(image)
            .getPalette()
            .then(palette => {
                setState({
                    isLoading: false,
                    error: false,
                    palette: palette,
                })
            })
            .catch(err => {
                setState({
                    isLoading: false,
                    error: true,
                    palette: null,
                })
            })
    }

    useEffect(() => {
        getColor(props.image)
    }, [props.image])

    const style = {
        backgroundColor:
            (state.palette &&
                state.palette[props.profile] &&
                rgbToString(state.palette[props.profile]._rgb)) ||
            defaultBackgroundColor,
        color: defaultColor,
        display: "flex",
        justifyContent: "center",
        alignItems: "center",
    }

    return (
        <Frame size="100%" style={style}>
            {state.isLoading && "Loading…"}
            {props.image === "" && "Choose an image"}
            {state.error && props.image && "Error"}
            {state.palette &&
                !state.palette[props.profile] &&
                "Unavailable color profile"}
        </Frame>
    )
}

Palette.defaultProps = {
    image: "",
    profile: "Vibrant",
}

addPropertyControls(Palette, {
    image: {
        type: ControlType.Image,
        title: "Image",
    },
    profile: {
        type: ControlType.Enum,
        options: [
            "LightVibrant",
            "Vibrant",
            "DarkVibrant",
            "LightMuted",
            "Muted",
            "DarkMuted",
        ],
        title: "Color profile",
        hidden(props) {
            return props.image === ""
        },
    },
})
